### ISSUE TITLE
nanoseconds time stamps auto rounding off to microseconds in grafana - elasticsearch queries

### ISSUE THREAD

**Issue by vmanju123** [grafana/grafana#23408]
**What happened**:
we have a field in our ES database called nanoTS, long integer basically a number representing nano seconds Eg: 1586266796413334123, when we query this in grafana, it automatically rounds it to nearest microseconds Eg: 1586266796413334000, the same query in Kibana provides accurate number as is in database, we are using logs based query to get entire index into a table panel

**What you expected to happen**:
No rounding off or if there is a option to turn off rounding

**How to reproduce it (as minimally and precisely as possible)**:
add a field in es db for long number give numbers as provided above

**Anything else we need to know?**:

**Environment**:
- Grafana version: 6.5.2
- Data source type & version: ElastisSearch 7.0+
- OS Grafana is installed on: Ubuntu 18.04
- User OS & Browser: Windows/Chrome
- Grafana plugins: Out of box plugins
- Others:

**Comment by aocenas** [comment]
Hey @vmanju123 issue is JS by default does not handle big integers and there is no special handling of that right now. I assume it is your primary timestamp but if not you could workaround this by using it as string which at least would give you accurate display and sorting.

**Comment by vmanju123** [comment]
Its not our primary time stamp field, but additional field, when you say string, you meant on es database change the field from long to string ?

**Comment by aocenas** [comment]
Yes unfortunately that would have to be changed in elastic. The conversion to number in Grafana and thus loss of precision happens during the response parsing, so it's not possible to fix that with formatting later on.

**Comment by Estrax** [comment]
@aocenas do we need any additional work there, or can we close this issue?

**Comment by gabor** [comment]
@kylebrandt do you know maybe, is there a plan in Grafana to have a way to store very large integers (larger than `2^^53-1`) in a dataframe-field? because currently numbers-in-a-field are represented as javascript-numbers in the javascript parts of Grafana, and those numbers break down above `2^^53-1` ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER ).

**Comment by gabor** [comment]
yeah, i assume in the backend-code this all works just fine (large intergers should be OK in `go`... i think)..

**Comment by leeoniya** [comment]
hmm, actually this will only let us pass these values through the json dataframe. they get decoded automatically into field values, where this would break down again. we'll need to figure out a way to skip this auto-decoding into full field values. tricky :(

**Comment by gabor** [comment]
@leeoniya thanks for the feedback. but, i think in this specific case, it's not important that it is a timestamp, it is just a large integer-value (i mean, these values are not in the "timestamp" field of the dataframe, there are just a numeric-field in elasticsearch)

**Comment by leeoniya** [comment]
so there are two issues here. the first is the limit on integer size. we can detect this and encode as needed to avoid precision loss.

the second one is that it sounds like the intent is to treat them as timestamps. i dont see how to square that circle without giving users explicit settings in the datasource UI to set specific numeric fields's `type` as `time`. but i'm just edu guessing here.

**Comment by gabor** [comment]
> so there are two issues here. the first is the limit on integer size. we can detect this and encode as needed to avoid precision loss.

could you describe this part in more detail?

> the second is the intent is to treat them as timestamps. i dont see how to square that circle without giving users explicit settings in the datasource UI to set specific numeric field's `type` as `time` or `time_ns`. but i'm just edu guessing here.

i think we can ignore this problem. i mean, the original issue here is about showing large-integers without deforming those large-integers (because of javascript.MAX_SAFE_INT). there is no need to do anything "timestamp-related", afaik.
